### PR TITLE
Fix session date parse issue

### PR DESCRIPTION
**Pull request by khroolick**
Fixed SessionId not sent to appinsight due to miliseconds added to a long date by JS appinsights library

### CODE REVIEW

**Comment by khroolick** [comment]
@Yonisha take a look at this please if you are still around 
@dhaval24 this one is a blocker for us

**Comment by dhaval24** [comment]
@khroolick Thank you very much for your contribution. Is this PR a fix for #537 ? As you said this one is blocker for you, does it mean you are not able to use AI java SDK in production without this fix?

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java [hunk 1]
```diff
@@ -191,8 +190,10 @@ private void parseCookie(Cookie cookie) throws Exception {
      */
     private Date parseDateWithBackwardCompatibility(String dateStr) throws ParseException {
         try {
-            return new Date(Long.parseLong(dateStr));
-        } catch (NumberFormatException e) {
+            Scanner s = new Scanner(dateStr);
+            s.useDelimiter("\\.");
```
Comment (dhaval24) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: Is the date string always going to have separators as "." ? Can date string ever be in format like 01/01/2018 ? If that is the case then this logic will not work. We might need to consider all seperators.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java [hunk 2]
```diff
@@ -22,8 +22,7 @@
 package com.microsoft.applicationinsights.web.internal.cookies;
 
 import java.text.ParseException;
-import java.util.Calendar;
-import java.util.Date;
+import java.util.*;
```
Comment (dhaval24) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: Please do not use * imports. Edit it to add individual import statements,

**Comment by gondar** [comment]
We don't have sessionId without it. This renders 70% of our data useless. We cannot do proper funnel analysis nor understand why given POST resulted in slow response (in order to get meaningful data we must correlate it with previous requests).
SessionId is not something that we can overlook when sending data to appInsights, it's a must for us to do analysis required by our management. :-D
Hope that helps.

**Comment by SergeyKanzhelev** [comment]
Date parsing is not required any longer. The field representing it was deprecated.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java (hunk 1 above)
Comment (gondar) [reply web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: @dhaval24 can you please check what AppInsights Javascript library sends you? I mean this is contract between to microsoft appinsights codebases one for JS and one for Java. I didn't find any documentation about it (no surprise as this is internal concern). I am sure you can get in touch with relevant team? 

Note that this bug was introduced over a year ago when js team started sending ms in cookie. So this is not about date (date was implementation used even earlier and you have code for backwards compatibility below), now it's format is <long_for_datetime>.<miliseconds> from looking at JS code https://github.com/Microsoft/ApplicationInsights-JS/blob/cac7b363a7f59147adff510f698bc3f4943f2ab0/JavaScript/JavaScriptSDK/ajax/ajaxUtils.ts#L36 (and we see that being a problem in our test env). Is there some ms spec on how ai_session cookie is meant to be?

**Comment by dhaval24** [comment]
@SergeyKanzhelev thank you very much for this. I believe with the new schema we have deprecated the SessionStateData  data structure and therefore none of that gets serialized, and sent to ingestion if I am correct?

**Comment by gondar** [comment]
This is not fix for #536 . This issue is to get sessionId working in situation where we have both appinsights JS in FE and appinsights Java in backend. Java cookie parsing logic is not compatible with JS cookie creation :-D.

**Comment by SergeyKanzhelev** [comment]
You can send SessinoStateData and session acquisition time. But both will be stripped out in our endpoint.

**Comment by gondar** [comment]
We did end to end test in our dev environment. To reproduce issue:
1. put newest JS appinsights snippet on a page
2. put current Java in your java application in standard setup (logging requests via webfilter as described in getting started guide)
3. send any requests to your application from any modern browser (ff, chrome)
4. go to appinsights portal and do following query:
requests
| where isnotempty(sessionId)
| summarize count() by sessionId

the result will be empty, whereas we should see count of requests per each session

:-)
Hope that helps.

**Comment by gondar** [comment]
@SergeyKanzhelev I agree it's not used for anything :-D. It actually duplicates expiry field in standard http.

**Comment by gondar** [comment]
@SergeyKanzhelev so you keep just sessionId and strip all after? @dhaval24 what do u think?

**Comment by SergeyKanzhelev** [comment]
@gondar yes. The code in .NET looks strange https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/47d18691f836ca6509ad2f5014c297930b380685/Src/Web/Web.Shared.Net/SessionTelemetryInitializer.cs#L19-L38 as it also used to have parsing of the date. Now just take a `substring(0, indexOf("|"))`

**Comment by khroolick** [comment]
@dhaval24 I've removed all this unused code altogether so there is nothing to cover with E2E test anymore :-)

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java (hunk 2 above)
Comment (khroolick) [reply web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: fixed

**Comment by dhaval24** [comment]
@khroolick thank you very much for this. @SergeyKanzhelev  could you please review this changes, as I am not very familier with this part of the code and also probably resembles more the schema. We can take this change then as a part of next Beta release after the stable.

**Comment by gondar** [comment]
Thanks all!
I would just point out that this is bug fix, not enhacement. Not having
sessionid in telemetry is serious limitation of data quality. Also not
working well with js cookie seems more a bug than anything.
Therefore my advise would be to apply it asap rather then with next
release, but its your choice. We decided to use our own fork with fix until
it is sorted.

27.01.2018 19:32 "Dhaval Doshi" <notifications@github.com> napisał(a):

**Comment by dhaval24** [comment]
@gondar appreciate your genuine feedback. I have updated the tag from enhancement to bug. We would discuss around taking this as a part of stable release coming next once @SergeyKanzhelev reviews this PR in depth. I appreciate you taking the pain to use your own fork for the time being. 

@grlima and @littleaj the issue which is described here is for sure a bug which is not desirable for most of the people using JS + Java SDK together. Please have a look and review this.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/HttpCookieFactory.java [hunk 3]
```diff
@@ -46,12 +45,9 @@
      */
     public static Cookie generateSessionHttpCookie(
```
Comment (SergeyKanzhelev) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/HttpCookieFactory.java]: So java SDK has a feature to generate session ID cookie if it wasn't set by JavaScript? How does it work around the problems mentioned here http://apmtips.com/blog/2016/07/09/tracking-users-in-api-apps/?

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java [hunk 4]
```diff
@@ -34,7 +30,6 @@
 
     // region Consts
 
-    private static final int UPDATE_TIMEOUT_IN_MINUTES = 5;
     public static final String COOKIE_NAME = "ai_session";
     public static final int SESSION_DEFAULT_EXPIRATION_TIMEOUT_IN_MINUTES = 30;
```
Comment (SergeyKanzhelev) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: do you need this variable any more?

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java [hunk 5]
```diff
@@ -103,99 +72,18 @@ public String getSessionId() {
...
-     */
-    private Date parseDateWithBackwardCompatibility(String dateStr) throws ParseException {
-        try {
-            return new Date(Long.parseLong(dateStr));
-        } catch (NumberFormatException e) {
-            return DateTimeUtils.parseRoundTripDateString(dateStr);
-        }
+    private static String parseCookie(Cookie cookie) {
+        String[] split = cookie.getValue().split(RAW_COOKIE_SPLIT_DELIMITER, 2);
+        return split[0];
```
Comment (SergeyKanzhelev) [inline web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: substring will work faster

**Comment by dhaval24** [comment]
@khroolick , Sergey has approved your PR but left few comments. Could you please take time to clean up some of the things he mentioned and also respond to one of the questions he asked. Thanks again for the contribution

**Comment by SergeyKanzhelev** [comment]
@dhaval24 I think the question is more for you and @grlima. You'd need to discuss this SDK behavior with usage team to make sure it will not screw up any numbers for them.

**Comment by dhaval24** [comment]
@SergeyKanzhelev thanks. Got it, we will discuss this.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java (hunk 5 above)
Comment (khroolick) [reply web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/SessionCookie.java]: I agree 
I will change it to use substring

**Comment by dhaval24** [comment]
@khroolick thank you for updating the PR. By the way it has merge conflicts with master. Could you please resolve the conflict so it is ready. @grlima and @littleaj could you please review and sign off this one.

File: web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/HttpCookieFactory.java (hunk 3 above)
Comment (grlima) [reply web/src/main/java/com/microsoft/applicationinsights/web/internal/cookies/HttpCookieFactory.java]: This method does seem to be used anywhere but unit tests. As a matter of fact, this factory class is not used at all. We should remove it. Created #577

**Comment by gondar** [comment]
@SergeyKanzhelev can you please notify relevant PMs about this bug. I find this dragging on 

@grlima How much time it will take you to verify? It took us 1h to spot that issue and 1h to fix it. I presume with dev env for app insights java you can spot it and see that it resolves it in 15 minutes, right? If not do you want to do middle of night paring session where I would show you how it works?

**Comment by SergeyKanzhelev** [comment]
CC: @nikmd23

**Comment by dhaval24** [comment]
@grlima this PR has a lot of public methods getting removed / getting refactored in signature also. Before you merge this, please update the readme to reflect this removals and  refactoring.

**Comment by grlima** [comment]
These changes will be merged with #622, which includes additional cleanup and updates to the change log.

**Pull request closed without merging**